src/entry-client.js

```diff
@@ -117,25 +124,23 @@ export default function initializeClient(createApp, clientOpts) {
         // Approach based on:
         //   https://ssr.vuejs.org/en/data.html#client-data-fetching
         router.beforeResolve((to, from, next) => {
-            // For simplicity, since we aren't using nested routes or anything fancy,
-            // we will just always call fetchData on the new components.  If we try to
-            // route to the same exact route, it shouldn't even fire the beforeResolve.
-            // And if we are routing to the same component with new params, then we
-            // likely want to be calling fetchData again.  If this proves to be too
-            // loose of an approach, a comprehensive approach is available at:
-            //   https://ssr.vuejs.org/en/data.html#client-data-fetching
-            const components = router.getMatchedComponents(to);
-            const fetchData = c => c.fetchData && c.fetchData({
-                app,
-                route: to,
-                router,
-                store,
-                from,
-            });
+            const routeUpdateStr = `${from.fullPath} -> ${to.fullPath}`;
+            const fetchDataArgs = { app, route: to, router, store, from };
+            const fetchData = c => isFunction(c.fetchData) && c.fetchData(fetchDataArgs);
+            const components = router.getMatchedComponents(to)
+                .filter(c => !shouldIgnoreRouteUpdate(c, fetchDataArgs));
+
+            // Short circuit if none of our components need to process the route update
+            if (components.length === 0) {
+                opts.logger.debug(`Ignoring route update ${routeUpdateStr}`);
+                return next();
+            }
+
+            opts.logger.debug(`Running middleware/fetchData for route update ${routeUpdateStr}`);
             return Promise.resolve()
-                .then(() => opts.middleware(to, from, store))
+                .then(() => opts.middleware(to, from, store, app))
                 .then(() => Promise.all(components.map(fetchData)))
-                .then(() => opts.postMiddleware(to, from, store))
+                .then(() => opts.postMiddleware(to, from, store, app))
```

Comment:
Updated to provide the `app` to the client middlewares as well as this was required for some of the restructuring this opened up in SlipStream.